plugins/commhistory/src/commhistoryplugin.cpp

```diff
@@ -0,0 +1,218 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "commhistoryplugin.h"
+
+#include <common.h>
+#include <voicecallmanagerinterface.h>
+#include <abstractvoicecallhandler.h>
+
+#include <CommHistory/EventModel>
+#include <CommHistory/Event>
+
+#include <QTimer>
+
+static QString activeCallProp = QLatin1String("activeCall");
+class CommHistoryPlugin::Private
+{
+public:
+    Private()
+    {
+        m_timer.setInterval(5 * 60000);
+    }
+
+    VoiceCallManagerInterface *m_manager = nullptr;
+    QMap<QString, CommHistory::Event> m_calls;
+    QTimer m_timer;
+    CommHistory::EventModel m_eventModel;
+
+    void newVoiceCall(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event event;
+        event.setType(CommHistory::Event::CallEvent);
+        event.setLocalUid(handler.provider()->providerId());
+        event.setRecipients(CommHistory::Recipient(event.localUid(),
+                                                   handler.lineId()));
+        event.setSubscriberIdentity(handler.subscriberId());
+        event.setStartTime(QDateTime::currentDateTime());
+        event.setEndTime(event.startTime());
+        event.setIsEmergencyCall(handler.isEmergency());
+        event.setDirection(handler.isIncoming() ? CommHistory::Event::Inbound : CommHistory::Event::Outbound);
+
+        if (m_calls.isEmpty()) {
+            m_timer.start();
+        }
+        storeCall(&event);
+
+        m_calls.insert(handler.handlerId(), event);
+    }
+
+    void onVoiceCallStatusChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+
+        switch (handler.status()) {
+        case AbstractVoiceCallHandler::STATUS_ACTIVE:
+            event.setStartTime(QDateTime::currentDateTime());
+            event.setEndTime(event.startTime());
+            event.setExtraProperty(activeCallProp, true);
+            storeCall(&event);
+            break;
+        case AbstractVoiceCallHandler::STATUS_DISCONNECTED:
+            event.setEndTime(event.startTime().addSecs(handler.duration()));
+            // No need to save the changes now, the voiceCallEnded will do it.
+            break;
+        default:
+            break;
+        }
+    }
+
+    void onVoiceCallEmergencyChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+        event.setIsEmergencyCall(handler.isEmergency());
+        storeCall(&event);
+    }
+
+    void onVoiceCallDurationChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+        event.setEndTime(event.startTime().addSecs(handler.duration()));
+    }
+
+    void storeCall(CommHistory::Event *event)
+    {
+        if (!event->isValid() && !m_eventModel.addEvent(*event)) {
```

Comment:
This was so that event is not valid until added to event, right? Looks funny as in ensuring only invalid events get added.

Reply:
At first I implemented it as :
```
if (event->id() == -1 && !model.addEvent(*event))
```

and then, I noticed from `libcommhistory/src/event.cpp` that `Event::isValid()` is implemented as `return id() != -1`, so I changed my first implementation to use this method instead. I agree that the name is a bit confusing. A `isStored()` would have reflected better its implementation.

If you think it's too misleading, I can go back to my first implementation.

Reply:
There's at least also declarative / DraftMessage with writeable eventId property. If someone sets an id there, it wouldn't necessarily mean it's stored yet (?). 

Maybe this is fine but could just have a comment about stored events being considered valid.

Reply:
I've added a comment mentioning that event becomes valid after being added to a model.